### PR TITLE
Stamp Redeem Functionality 

### PR DESCRIPTION
**Pull request by Patrickrgt**
# Title: Implement and Integrate Stamp Redemption Functionality with Backend APIs

## Overview:

This PR introduces a feature allowing users to redeem rewards (in the form of discounts) based on the number of Polygon NFTs (or "stamps") they hold. The feature is integrated with an API deployed on Heroku for the management of these discounts. 

The API, which can be found at [https://aiko-api-demo-2bb37db66bbe.herokuapp.com/](https://aiko-api-demo-2bb37db66bbe.herokuapp.com/), is used to fetch, generate, and assign discounts. The details of the API and the new functionalities are provided below.

## Description:

The functionality is designed to incentivize users by offering them rewards based on the quantity of stamps they hold. The reward quantity and coupon code are encapsulated into a single encrypted string. A 'Redeem' button is added which, upon clicking, redirects the user to a Shopify link. The Shopify page retrieves the encrypted code from the URL slug, decrypts it, and displays available rewards for the user to add to their cart and checkout.

## New Functionality:

- `toBase64Url(base64: string)`: Converts a base64 string to URL-safe format.
  
- `fromBase64Url(base64url: string)`: Converts a URL-safe base64 string back to the standard format.
  
- `encrypt(object: Record<string, unknown>, key: string)`: Encrypts an object with a key, and converts the output into a URL-safe base64 string.

- `decrypt(ciphertext: string, key: string)`: Decrypts an encrypted string with a key, converting it back to the original JSON object.

- `isValidDecryptedObject(object: any)`: Checks if the provided object is a valid DecryptedObject.

- `generateRandomCouponCode(length: number)`: Generates a random alphanumeric string of the specified length.

- `fetchDiscount(existsId: number)`: Fetches a discount based on an existing ID from the backend.

- `useEffect(() => {...}, [encryptedObject, key])`: React hook which triggers every time `encryptedObject` or `key` undergo changes.

- `postDiscount()`: Asynchronous function that fetches all existing discounts, checks for a discount associated with the user's wallet, and either fetches its details or generates a new one.

- `handleButtonClick()`: Triggers `postDiscount` if the user has at least 3 stamps, or disables the button otherwise.

## Redemption Criteria:

- Users holding 3-5 stamps are eligible for 1 reward.
- Users holding 6-8 stamps are eligible for 2 rewards.
- Users holding 9-12 stamps are eligible for 3 rewards.

## How it Works:

1. The system checks for the number of stamps (Polygon NFTs) held by a user.
2. If the user falls into one of the above tiers, the system will generate a coupon for the corresponding number of rewards.
3. The coupon code and reward quantity are then encrypted into a string.
4. Once the 'Redeem' button is clicked, a new window/tab opens redirecting the user to a specific Shopify link with the encrypted code appended to the URL slug.
5. The Shopify page then decrypts the URL slug to reveal the coupon code and reward quantity, allowing the user to add the reward(s) to their cart and proceed to checkout.

## API:

- `GET /get-discounts`

   This route interacts with the Shopify API to fetch price rules (discounts). It sends a GET request to the Shopify Price Rules API endpoint and returns the data as a JSON response if successful. In case of an error, it returns a corresponding error message.

- `GET /get-stock/:inventory_item_id`

   This route is responsible for retrieving stock levels based on a specific inventory item ID, which is provided as a URL parameter. It sends a GET request to the Shopify Inventory Levels API endpoint and returns the data as a JSON response if successful. In case of an error, it returns a corresponding error message.

- `GET /get-discount/:id`

   This route fetches a specific discount based on a provided ID. It sends a GET request to the Shopify Price Rules and Discount Codes API endpoint using the provided ID in the URL of the request. It returns the fetched data as a JSON response if successful. In case of an error, it returns a corresponding error message.

- `POST /generate-discount`

   This route generates a new discount code. It accepts a request body with the `basicCodeDiscount` value and constructs a GraphQL mutation query to create a new discount code. The mutation is sent to the Shopify GraphQL API endpoint. If successful and no user errors are found, it returns the newly created discount code. If there are user errors or other exceptions, it returns a corresponding error message.

### CODE REVIEW

File: src/components/userpanel/StampRedeem.tsx [hunk 1]
```diff
@@ -18,27 +25,176 @@ const ShippingText = styled.span`
...
+  const stampsHeld = useSelector(selectStampsHeld);
+
+  const fetchDiscount = async (existsId: number) => {
+    try {
+      const response = await fetch(`${aikoAPI}/get-discount/${existsId}`);
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+      const data = await response.json();
+      // setCode(data.discount_codes[0].code);
```
Comment (chase-manning) [inline src/components/userpanel/StampRedeem.tsx:63]: Can probably remove the comment and logging

File: src/components/userpanel/StampRedeem.tsx [hunk 2]
```diff
@@ -18,27 +25,176 @@ const ShippingText = styled.span`
...
+      } else {
+        console.error(
+          "Decrypted object does not have the expected shape:",
+          object
+        );
+      }
+    }
+  }, [encryptedObject, key]);
+
+  async function postDiscount() {
```
Comment (chase-manning) [inline src/components/userpanel/StampRedeem.tsx:91]: Might be worth moving all of this logic outside of the component and into some sort of service or separate file 😄

File: src/components/userpanel/StampRedeem.tsx (hunk 1 above)
Comment (Patrickrgt) [reply src/components/userpanel/StampRedeem.tsx]: Done! Thank you!

File: src/components/userpanel/StampRedeem.tsx (hunk 2 above)
Comment (Patrickrgt) [reply src/components/userpanel/StampRedeem.tsx]: Done! Thank you!

**Review by chase-manning — APPROVED** [review]

**Pull request merged**